pinot-core/src/main/java/com/linkedin/pinot/core/transport/netty/ServerChannels.java

```diff
@@ -0,0 +1,98 @@
+
+import com.linkedin.pinot.common.metrics.BrokerMetrics;
+import com.linkedin.pinot.common.request.InstanceRequest;
+import com.linkedin.pinot.core.transport.QueryRouter;
+import com.linkedin.pinot.core.transport.Server;
+import com.linkedin.pinot.core.transport.netty.handler.DataTableHandler;
+import io.netty.bootstrap.Bootstrap;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
+import io.netty.handler.codec.LengthFieldPrepender;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.thrift.TSerializer;
+import org.apache.thrift.protocol.TCompactProtocol;
+
+
+/**
+ * The {@code ServerChannels} class manages the channels between broker to all the connected servers.
+ * <p>There is only one channel between the broker and each connected server (we count OFFLINE and REALTIME as different
+ * servers)
+ */
+@ThreadSafe
+public class ServerChannels {
+  private final QueryRouter _queryRouter;
+  private final BrokerMetrics _brokerMetrics;
+  private final ConcurrentHashMap<Server, ServerChannel> _serverToChannelMap = new ConcurrentHashMap<>();
+  private final EventLoopGroup _eventLoopGroup = new NioEventLoopGroup();
+
+  public ServerChannels(QueryRouter queryRouter, BrokerMetrics brokerMetrics) {
+    _queryRouter = queryRouter;
+    _brokerMetrics = brokerMetrics;
+  }
+
+  public void sendRequest(Server server, InstanceRequest instanceRequest) throws Exception {
+    _serverToChannelMap.computeIfAbsent(server, ServerChannel::new).sendRequest(instanceRequest);
+  }
+
+  @ThreadSafe
+  private class ServerChannel {
+    final TSerializer _serializer = new TSerializer(new TCompactProtocol.Factory());
+    final Server _server;
+    final Bootstrap _bootstrap;
+    Channel _channel;
+
+    ServerChannel(Server server) {
+      _server = server;
+      _bootstrap = new Bootstrap().remoteAddress(server.getHostName(), server.getPort())
+          .group(_eventLoopGroup)
+          .channel(NioSocketChannel.class)
+          .option(ChannelOption.SO_KEEPALIVE, true)
+          .option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
+          .handler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            protected void initChannel(SocketChannel ch) {
+              ch.pipeline()
+                  .addLast(new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 0, Integer.BYTES, 0, Integer.BYTES),
+                      new LengthFieldPrepender(Integer.BYTES),
+                      new DataTableHandler(_queryRouter, _server, _brokerMetrics));
```

Comment:
I don't want to add more threads to the broker. If this becomes a bottleneck, we can change it later.

Reply:
In that case, please add some comments there, thanks.

Reply:
Actually this is better than the previous implementation where all the de-serialization happens in a single thread. Any reason why you think this could cause problem?